streaming/base/spanner.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Mapping of global sample index to shard and relative sample index."""
+
+from typing import Tuple
+
+import numpy as np
+from numpy.typing import NDArray
+
+
+class Spanner:
+    """Given a list of shards, construct a mapping of global index to shard and relative index.
+
+    Args:
+        shard_sizes (NDArray[np.int64]): Number of samples in each shard.
+        span_size (int): Size of the divisions of the sample space. Defaults to ``1 << 10``.
+    """
+
+    def __init__(self, shard_sizes: NDArray[np.int64], span_size: int = 1 << 10) -> None:
+        self.shard_sizes = shard_sizes
+        self.span_size = span_size
+        self.size = sum(shard_sizes)
+        self.shard_bounds = np.concatenate([np.zeros(1, np.int64), shard_sizes.cumsum()])
+
+        overflow = self.size % span_size
+        underflow = span_size - overflow if overflow else 0
+        self.shard_sizes[-1] += underflow
+
+        sample_shards = np.repeat(np.arange(len(shard_sizes)), self.shard_sizes)
+        sample_shards = sample_shards.reshape(-1, span_size)
```

Comment:
I'm not sure I understand

If the number of samples isn't divisible by span size (the base case), it just pads the last shard to round up, then unpads after it's done. So it will always be reshapeable